File: ClientApp/src/app/services/employee.service.ts
import { HttpClient } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { Employee } from "../models/employee.model";
import { ERole } from "./../models/eRole.model";

const employeesUrl = "/api/employees";

@Injectable()
export class EmployeeService {
  constructor(private http: HttpClient) {}

  getEmployee(id: string) {
    return this.http.get<Employee>(`${employeesUrl}/${id}`);
  }

  getEmployees(managerId: string) {
    let url = employeesUrl;
    if (managerId) url = `${employeesUrl}?managerId=${managerId}`;

    return this.http.get<Employee[]>(url);
  }

  getManagers(excludeEmployeeId: string) {
    let url = `${employeesUrl}/managers/${excludeEmployeeId}`;
    return this.http.get<Employee[]>(url);
  }

  getAllRoles() {
    const url = `${employeesUrl}/roles`;
    return this.http.get<ERole[]>(url);
  }

  getRolesByEmployeeId(empId: string) {
    const url = `${employeesUrl}/roles/${empId}`;
    return this.http.get<ERole[]>(url);
  }

  createEmployee(emp: Employee) {
    let data = {
      employeeId: emp.employeeId,
      firstName: emp.firstName,
      lastName: emp.lastName,
      managerId: emp.managerId,
      employeeRoles: [],
    };

    if (emp.eRoles && emp.eRoles.length > 0) {
      let roles = [];
      for (let rId of emp.eRoles)
        roles.push({ employeeId: emp.employeeId, eRoleId: rId.eRoleId });

      data.employeeRoles = roles;
    }

    return this.http.post<number>(employeesUrl, data);
  }

  updateEmployee(emp: Employee) {
    let data = {
      employeeId: emp.employeeId,
      firstName: emp.firstName,
      lastName: emp.lastName,
      managerId: emp.managerId,
      employeeRoles: [],
    };

    if (emp.eRoles && emp.eRoles.length > 0) {
      let roles = [];
      for (let rId of emp.eRoles)
        roles.push({ employeeId: emp.employeeId, eRoleId: rId.eRoleId });

      data.employeeRoles = roles;
    }

    return this.http.put<number>(`${employeesUrl}/${emp.employeeId}`, data);
  }
}

File: SSEPS.Service/Implementation/EmployeeService.cs
using SSEPS.Domain;
using SSEPS.Repository;
using System.Collections.Generic;

namespace SSEPS.Service
{
    public class EmployeeService: IEmployeeService
    {
        private IEmployeeRepository repository;

        public EmployeeService(IEmployeeRepository repo)
        {
            repository = repo;
        }

        public IEnumerable<Employee> GetEmployees(string managerId = "")
        {
            var employees = repository.GetEmployees(managerId);
            
            foreach (var emp in employees)
            {
                if (emp.Manager != null)
                    emp.Manager.Manager = null;
            }

            return employees;
        }

        public IEnumerable<Employee> GetAllManagers(string excludedEmployeeId)
        {
            var managers = repository.GetAllManagers(excludedEmployeeId);
            return managers;
        }

        public Employee GetEmployee(string id)
        {
            var employee = repository.GetEmployee(id);
            
            if (employee?.Manager != null)
                employee.Manager.Manager = null;

            return employee;
        }

        public string CreateEmployee(Employee employee)
        {
            var employeeId = repository.CreateEmployee(employee);
            return employeeId;
        }

        public Employee UpdateEmployee(string id, Employee employee)
        {
            var dbEmployee = repository.GetEmployee(id);
            if (dbEmployee == null)
                return null;

            repository.RemoveRoles(employee.EmployeeId);

            dbEmployee.FirstName = employee.FirstName;
            dbEmployee.LastName = employee.LastName;
            dbEmployee.ManagerId = employee.ManagerId;
            dbEmployee.EmployeeRoles = employee.EmployeeRoles;

            return repository.UpdateEmployee(dbEmployee);
        }

        public void DeleteEmployee(string id)
        {
            repository.DeleteEmployee(id);
        }

        public IEnumerable<ERole> GetAllRoles()
        {
            return repository.GetAllRoles();
        }

        public IEnumerable<ERole> GetRoles(string employeeId)
        {
            return repository.GetRoles(employeeId);
        }

        public void RemoveRoles(string employeeId)
        {
            repository.RemoveRoles(employeeId);
        }
    }
}
